Data race in ipam.TestSpaceRequest

In a build of the `--ipam-seed` branch - believed to be unrelated to the changes on that branch: https://circleci.com/gh/weaveworks/weave/4958

```
git submodule update --init
sudo -E docker run  -ti \
	    -v /home/ubuntu/src/github.com/weaveworks/weave:/go/src/github.com/weaveworks/weave \
		-v /home/ubuntu/src/github.com/weaveworks/weave/.pkg:/go/pkg \
		-e GOARCH -e GOOS -e CIRCLECI -e CIRCLE_BUILD_NUM -e CIRCLE_NODE_TOTAL -e CIRCLE_NODE_INDEX -e COVERDIR -e SLOW \
		weaveworks/weavebuild COVERAGE= WEAVE_VERSION=git-488d2f85c12c tests
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
make: Entering directory '/go/src/github.com/weaveworks/weave'
./tools/test -no-go-get
./ipam ./ipam/paxos ./net
INFO: 2016/03/14 17:54:33.224358 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.228080 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.228308 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.337403 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.337851 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.341069 [allocator 01:00:00:02:00:00] Claim 10.0.3.2/24 for b01df00d: address allocator still initializing; will try later.
INFO: 2016/03/14 17:54:33.361307 [allocator 01:00:00:02:00:00] Address 10.0.4.2/24 claimed by abcdef - not in our range
INFO: 2016/03/14 17:54:33.363092 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.363345 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.474388 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.575802 [allocator 01:00:00:01:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.576357 [allocator 01:00:00:01:00:00] Shutdown
INFO: 2016/03/14 17:54:33.577058 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.577441 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.577815 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.624015 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.624426 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.637501 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.638206 [allocator 01:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.638727 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.639159 [allocator 03:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:33.639580 [allocator 04:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:54.036961 [allocator 00:00:00:02:00:00] Initialising via deferred consensus
==================
WARNING: DATA RACE
Write by goroutine 72:
  github.com/weaveworks/weave/ipam/ring.(*Ring).Merge()
      github.com/weaveworks/weave/ipam/ring/_obj/ring.go:334 +0x125e
  github.com/weaveworks/weave/ipam.(*Allocator).update()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:859 +0x899
  github.com/weaveworks/weave/ipam.(*Allocator).OnGossipBroadcast.func1()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:591 +0xad
  github.com/weaveworks/weave/ipam.(*Allocator).actorLoop()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:698 +0x1aa

Previous read by goroutine 70:
  github.com/weaveworks/weave/ipam/ring.(*Ring).checkInvariants()
      github.com/weaveworks/weave/ipam/ring/_obj/ring.go:99 +0x75f
  github.com/weaveworks/weave/ipam/ring.(*Ring).assertInvariants()
      github.com/weaveworks/weave/ipam/ring/_obj/ring.go:33 +0x4a
  github.com/weaveworks/weave/ipam/ring.(*Ring).OwnedRanges()
      github.com/weaveworks/weave/ipam/ring/_obj/ring.go:378 +0x7b
  github.com/weaveworks/weave/ipam.TestSpaceRequest()
      /go/src/github.com/weaveworks/weave/ipam/allocator_test.go:595 +0x510
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Goroutine 72 (running) created at:
  github.com/weaveworks/weave/ipam.(*Allocator).Start()
      github.com/weaveworks/weave/ipam/_test/_obj_test/allocator.go:108 +0x135
  github.com/weaveworks/weave/ipam.makeNetworkOfAllocators()
      /go/src/github.com/weaveworks/weave/ipam/testutils_test.go:214 +0x3cb
  github.com/weaveworks/weave/ipam.TestSpaceRequest()
      /go/src/github.com/weaveworks/weave/ipam/allocator_test.go:586 +0x54
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Goroutine 70 (running) created at:
  testing.RunTests()
      /tmp/workdir/go/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /tmp/workdir/go/src/testing/testing.go:494 +0xe4
  main.main()
      github.com/weaveworks/weave/ipam/_test/_testmain.go:186 +0x38d
==================
INFO: 2016/03/14 17:54:54.044634 [allocator 02:00:00:02:00:00] Initialising via deferred consensus
INFO: 2016/03/14 17:54:54.051640 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
INFO: 2016/03/14 17:54:54.064915 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
INFO: 2016/03/14 17:54:54.072675 [allocator 08:00:27:01:c3:9a] Initialising via deferred consensus
PASS
coverage: 64.3% of statements in github.com/weaveworks/weave/ipam, github.com/weaveworks/weave/common, github.com/weaveworks/weave/common/docker, github.com/weaveworks/weave/db, github.com/weaveworks/weave/ipam/paxos, github.com/weaveworks/weave/ipam/ring, github.com/weaveworks/weave/ipam/space, github.com/weaveworks/weave/net/address
Found 1 data race(s)
FAIL	github.com/weaveworks/weave/ipam	21.987s
ok  	github.com/weaveworks/weave/ipam/paxos	22.110s	coverage: 88.9% of statements in github.com/weaveworks/weave/ipam/paxos
ok  	github.com/weaveworks/weave/net	1.128s	coverage: 1.4% of statements in github.com/weaveworks/weave/net

Makefile:119: recipe for target 'tests' failed
make: *** [tests] Error 1
make: Leaving directory '/go/src/github.com/weaveworks/weave'

make: *** [tests] Error 2 cd $SRCDIR; COVERDIR=test/coverage make RM= tests returned exit code 2
```

--- a/ipam/allocator_test.go
+++ b/ipam/allocator_test.go
@@ -597,7 +597,7 @@ func TestSpaceRequest(t *testing.T) {
 	// free it again so the donation splits the range neatly
 	err = alloc1.Free(container1, addr)
 	require.Nil(t, err, "")
-	require.Equal(t, cidrRanges(universe), alloc1.ring.OwnedRanges(), "")
+	require.Equal(t, cidrRanges(universe), alloc1.OwnedRanges(), "")
 
 	// Start a new peer
 	alloc2, _ := makeAllocator("02:00:00:02:00:00", universe, 2)
@@ -607,8 +607,8 @@ func TestSpaceRequest(t *testing.T) {
 	alloc2.Allocate(container2, subnet, returnFalse)
 
 	// Test whether the universe has been split into two equal halves (GH #2009)
-	require.Equal(t, cidrRanges("10.32.0.0/13"), alloc1.ring.OwnedRanges(), "")
-	require.Equal(t, cidrRanges("10.40.0.0/13"), alloc2.ring.OwnedRanges(), "")
+	require.Equal(t, cidrRanges("10.32.0.0/13"), alloc1.OwnedRanges(), "")
+	require.Equal(t, cidrRanges("10.40.0.0/13"), alloc2.OwnedRanges(), "")
 }
 
 func cidrRanges(s string) []address.Range {

--- a/ipam/testutils_test.go
+++ b/ipam/testutils_test.go
@@ -175,6 +175,14 @@ func (alloc *Allocator) NumFreeAddresses(r address.Range) address.Count {
 	return <-resultChan
 }
 
+func (alloc *Allocator) OwnedRanges() (result []address.Range) {
+	resultChan := make(chan []address.Range)
+	alloc.actionChan <- func() {
+		resultChan <- alloc.ring.OwnedRanges()
+	}
+	return <-resultChan
+}
+
 // Check whether or not something was sent on a channel
 func AssertSent(t *testing.T, ch <-chan bool) {
 	timeout := time.After(10 * time.Second)

